{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "88057030",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import re \n",
    "from utils import * \n",
    "import os \n",
    "from src.ggkbase import * \n",
    "import glob\n",
    "import matplotlib as mpl \n",
    "from src.files import BLASTFileJSON, FASTAFile, InterProScanFileTSV, GenBankFile, fasta_get_genome_size\n",
    "import subprocess\n",
    "import shutil\n",
    "import networkx as nx\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a601d39d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Want to evaluate the likelihood of other potential hosts. Look for all other Methanoperedens in the bioreactors. \n",
    "# For some reason, it seems like things aren't appropriately-labeled in the coassembly. \n",
    "organism_info_df = list()\n",
    "for path in glob.glob('../data/ggkbase/*organism_info.tsv'):\n",
    "    organism_info_df.append(load_organism_info(path))\n",
    "organism_info_df = pd.concat(organism_info_df)\n",
    "\n",
    "mp_organism_info_df = organism_info_df[organism_info_df.taxonomy.str.contains('Methanoperedens') | organism_info_df.index.str.contains('Methanoperedens')].copy()\n",
    "bb_organism_info_df = organism_info_df[organism_info_df.index.str.contains('Black_Borg')].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26b1d380",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_fastani(input_dir:str='../data/ggkbase/contigs', output_dir:str='../data/ggkbase'):\n",
    "    input_list_path = os.path.join('.', 'fastani_input.txt')\n",
    "    output_path = os.path.join(output_dir, 'fastani.tsv')\n",
    "    input_paths = list(glob.glob(os.path.join(input_dir, '*')))\n",
    "    with open(input_list_path, 'w') as f:\n",
    "        f.write('\\n'.join(input_paths))\n",
    "    cmd = f'fastANI --ql {input_list_path} --rl {input_list_path} -o {output_path}'\n",
    "    subprocess.run(cmd, shell=True, check=True)\n",
    "    os.remove(input_list_path)\n",
    "# run_fastani()\n",
    "\n",
    "def load_fastani(path:str='../data/ggkbase/fastani.tsv'):\n",
    "    cols = ['query_genome_path', 'ref_genome_path', 'ani', 'n_aligned', 'n_fragments']\n",
    "    fastani_df = pd.read_csv(path, sep='\\t', names=cols)\n",
    "    fastani_df['query_genome_id'] = fastani_df.query_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "    fastani_df['ref_genome_id'] = fastani_df.ref_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "    return fastani_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26370a85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. separate Methanoperedens strains: 5\n"
     ]
    }
   ],
   "source": [
    "threshold = 99\n",
    "fastani_df = load_fastani()\n",
    "fastani_df = fastani_df[fastani_df.ani > threshold].copy()\n",
    "fastani_df = fastani_df[fastani_df.query_genome_id.isin(mp_organism_info_df.index.values)].copy()\n",
    "\n",
    "edges = list(zip(fastani_df.query_genome_id, fastani_df.ref_genome_id))\n",
    "graph = nx.Graph(edges)\n",
    "strains = list(nx.connected_components(graph))\n",
    "print('Num. separate Methanoperedens strains:', len(strains))\n",
    "\n",
    "strain_ids = {genome_id:f'mp_{i + 1}' for i, genome_ids in enumerate(strains) for genome_id in genome_ids}\n",
    "\n",
    "mp_organism_info_df['strain_id'] = mp_organism_info_df.index.map(strain_ids)\n",
    "# Use the biggest assembly as the representative. These are the ones used to create the id_to_ggkbase_name_map.\n",
    "mp_organism_info_df = mp_organism_info_df.sort_values('bin_length', ascending=False).drop_duplicates('strain_id')\n",
    "mp_organism_info_df = mp_organism_info_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "54e76427",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for id_, ggkbase_name in id_to_ggkbase_name_map.items():\n",
    "    # print(f'cp ../data/ggkbase/contigs/{ggkbase_name}.contigs.fa ../data/data/{id_}.fn')\n",
    "    subprocess.run(f'cp ../data/ggkbase/contigs/{ggkbase_name}.contigs.fa ../data/data/{id_}.fn', shell=True, check=True)\n",
    "    subprocess.run(f'cp ../data/ggkbase/genbank/{ggkbase_name}.gbk ../data/data/{id_}.gbk', check=True, shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e78bd395",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
