{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bf1723ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import * \n",
    "import os \n",
    "from src.ggkbase import * \n",
    "import glob\n",
    "import matplotlib as mpl \n",
    "from src.files import BLASTFileJSON, FASTAFile, InterProScanFileTSV, GenBankFile, fasta_get_genome_size\n",
    "import subprocess\n",
    "import networkx as nx\n",
    "import json\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Notebook for all the setup stuff. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "192abbde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Want to evaluate the likelihood of other potential hosts. Look for all other Methanoperedens in the bioreactors. \n",
    "# For some reason, it seems like things aren't appropriately-labeled in the coassembly. \n",
    "organism_info_df = list()\n",
    "for path in glob.glob('../data/ggkbase/*organism_info.tsv'):\n",
    "    organism_info_df.append(load_organism_info(path))\n",
    "organism_info_df = pd.concat(organism_info_df)\n",
    "mp_organism_info_df = organism_info_df[organism_info_df.taxonomy.str.contains('Methanoperedens') | organism_info_df.index.str.contains('Methanoperedens')].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "af01a67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_fastani(input_dir:str='../data/ggkbase/contigs', output_dir:str='../data/ggkbase'):\n",
    "    input_list_path = os.path.join('.', 'fastani_input.txt')\n",
    "    output_path = os.path.join(output_dir, 'fastani.tsv')\n",
    "    input_paths = list(glob.glob(os.path.join(input_dir, '*')))\n",
    "    with open(input_list_path, 'w') as f:\n",
    "        f.write('\\n'.join(input_paths))\n",
    "    cmd = f'fastANI --ql {input_list_path} --rl {input_list_path} -o {output_path}'\n",
    "    subprocess.run(cmd, shell=True, check=True)\n",
    "    os.remove(input_list_path)\n",
    "    \n",
    "# run_fastani()\n",
    "\n",
    "def load_fastani(path:str='../data/ggkbase/fastani.tsv'):\n",
    "    cols = ['query_genome_path', 'ref_genome_path', 'ani', 'n_aligned', 'n_fragments']\n",
    "    fastani_df = pd.read_csv(path, sep='\\t', names=cols)\n",
    "    fastani_df['query_genome_id'] = fastani_df.query_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "    fastani_df['ref_genome_id'] = fastani_df.ref_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "    return fastani_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6b7c0e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. separate Methanoperedens strains: 5\n"
     ]
    }
   ],
   "source": [
    "threshold = 99\n",
    "mp_genome_ids = mp_organism_info_df.index.tolist()\n",
    "\n",
    "fastani_df = load_fastani()\n",
    "fastani_df = fastani_df[fastani_df.ani > threshold].copy()\n",
    "fastani_df = fastani_df[fastani_df.query_genome_id.isin(mp_genome_ids) | fastani_df.ref_genome_id.isin(mp_genome_ids)].copy()\n",
    "\n",
    "edges = list(zip(fastani_df.query_genome_id, fastani_df.ref_genome_id))\n",
    "graph = nx.Graph(edges)\n",
    "strains = list(nx.connected_components(graph))\n",
    "print('Num. separate Methanoperedens strains:', len(strains))\n",
    "\n",
    "strain_ids = {genome_id:f'mp_{i + 1}' for i, genome_ids in enumerate(strains) for genome_id in genome_ids}\n",
    "\n",
    "mp_organism_info_df['strain_id'] = mp_organism_info_df.index.map(strain_ids)\n",
    "# Use the biggest assembly as the representative. These are the ones used to create the genome_id_to_ggkbase_id_map.\n",
    "mp_organism_info_df = mp_organism_info_df.sort_values('bin_length', ascending=False).drop_duplicates('strain_id')\n",
    "mp_organism_info_df = mp_organism_info_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a60a4a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_id_to_ggkbase_id_map = pd.read_csv('genome_metadata.csv').set_index('genome_id').ggkbase_id.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d02e894",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_contig_index = lambda gene_id : gene_id.split('.')[-1].split('_')[0]\n",
    "\n",
    "for genome_id, ggkbase_id in genome_id_to_ggkbase_id_map.items():\n",
    "    fasta_file = FASTAFile.from_file(f'../data/ggkbase/contigs/{ggkbase_id}.contigs.fa')\n",
    "    fasta_file.ids = [f'{genome_id}.{i + 1}' for i in range(len(fasta_file.ids))] # Rename the contigs\n",
    "    fasta_file.write(f'../data/data/{genome_id}.fn')\n",
    "\n",
    "    genbank_file = GenBankFile.from_file(f'../data/ggkbase/genbank/{ggkbase_id}.gbk')\n",
    "    genbank_file.df['gene_id'] = [f'{genome_id}.{gene_id}' for gene_id in genbank_file.df.gene_id] # Rename the gene IDs.\n",
    "    genbank_file.df['contig_id'] = [f'{genome_id}.{get_contig_index(gene_id)}' for gene_id in genbank_file.df.gene_id] # Rename the contig IDs.\n",
    "    subprocess.run(f'cp ../data/ggkbase/genbank/{ggkbase_id}.gbk ../data/data/{genome_id}.gbk', check=True, shell=True)\n",
    "    genbank_file.to_gff(f'../data/data/{genome_id}.gff')\n",
    "    genbank_file.to_fasta(f'../data/data/{genome_id}.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "19a619b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/data/all.fn', 'w') as f:\n",
    "    paths = [path for path in glob.glob('../data/data/*fn') if (os.path.basename(path) != 'all.fn')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read())\n",
    "        \n",
    "with open('../data/data/all.fa', 'w') as f:\n",
    "    paths = [path for path in glob.glob('../data/data/*fa') if (os.path.basename(path) != 'all.fa')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read())\n",
    "\n",
    "gff_header = '##gff-version  3\\n'\n",
    "with open('../data/data/all.gff', 'w') as f:\n",
    "    f.write(gff_header)\n",
    "    paths = [path for path in glob.glob('../data/data/*gff') if (os.path.basename(path) != 'all.gff')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read().replace(gff_header, ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a7d74003",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>ggkbase_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mp_4</td>\n",
       "      <td>SR-VP_11_27_2022_S1_80cm_Methanoperedens_44_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mp_1</td>\n",
       "      <td>SR-VP_05_06_2024_ck_bottom_Methanoperedens_44_47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mp_3</td>\n",
       "      <td>SR-VP_05_06_2024_N_top_Methanoperedens_44_14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mp_5</td>\n",
       "      <td>SR-VP_05_06_2024_N_top_Candidatus_Methanopered...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mp_2</td>\n",
       "      <td>SR-VP_05_06_2024_ck_bottom_Methanoperedens_41_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>jupiter_mini_borg_1</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>jupiter_mini_borg_2</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>jupiter_mini_borg_3</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>jupiter_mini_borg_4</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>jupiter_mini_borg_6</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>jupiter_mini_borg_7</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>jupiter_mini_borg_8</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>jupiter_mini_borg_9</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saturn_mini_borg_1</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_35_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saturn_mini_borg_2</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saturn_mini_borg_3</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>saturn_mini_borg_4</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>unclassified_mini_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_mini-Borg_reminsce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>unclassified_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_new_Borg_34_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>amethyst_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Amethyst_Borg_34_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>oxblood_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Oxblood_Borg_33_20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pink_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Pink_Borg_32_55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>purple_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Purple_Borg_33_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rose_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Rose_Borg_31_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>vermilion_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Vermilion_Borg_34_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>mercury_mini_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Mercury_mini-Borg_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>saturn_mini_borg_like</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Saturn_mini-Borg-l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ruby_borg_related</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Ruby-Borg-related_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>black_borg</td>\n",
       "      <td>SR-VP_05_06_2024_coassembly_Black_Borg_32_272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>linear_ece_19kb</td>\n",
       "      <td>Final_SR-VP_05_06_2024_coassembly_19kb_linear_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 genome_id                                         ggkbase_id\n",
       "0                     mp_4      SR-VP_11_27_2022_S1_80cm_Methanoperedens_44_5\n",
       "1                     mp_1   SR-VP_05_06_2024_ck_bottom_Methanoperedens_44_47\n",
       "2                     mp_3       SR-VP_05_06_2024_N_top_Methanoperedens_44_14\n",
       "3                     mp_5  SR-VP_05_06_2024_N_top_Candidatus_Methanopered...\n",
       "4                     mp_2   SR-VP_05_06_2024_ck_bottom_Methanoperedens_41_16\n",
       "5      jupiter_mini_borg_1  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "6      jupiter_mini_borg_2  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "7      jupiter_mini_borg_3  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "8      jupiter_mini_borg_4  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "9      jupiter_mini_borg_6  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "10     jupiter_mini_borg_7  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "11     jupiter_mini_borg_8  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "12     jupiter_mini_borg_9  SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_...\n",
       "13      saturn_mini_borg_1  SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_35_7\n",
       "14      saturn_mini_borg_2  SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_3...\n",
       "15      saturn_mini_borg_3  SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_3\n",
       "16      saturn_mini_borg_4  SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_4\n",
       "17  unclassified_mini_borg  SR-VP_05_06_2024_coassembly_mini-Borg_reminsce...\n",
       "18       unclassified_borg         SR-VP_05_06_2024_coassembly_new_Borg_34_11\n",
       "19           amethyst_borg     SR-VP_05_06_2024_coassembly_Amethyst_Borg_34_3\n",
       "20            oxblood_borg     SR-VP_05_06_2024_coassembly_Oxblood_Borg_33_20\n",
       "21               pink_borg        SR-VP_05_06_2024_coassembly_Pink_Borg_32_55\n",
       "22             purple_borg       SR-VP_05_06_2024_coassembly_Purple_Borg_33_3\n",
       "23               rose_borg         SR-VP_05_06_2024_coassembly_Rose_Borg_31_2\n",
       "24          vermilion_borg    SR-VP_05_06_2024_coassembly_Vermilion_Borg_34_8\n",
       "25       mercury_mini_borg  SR-VP_05_06_2024_coassembly_Mercury_mini-Borg_...\n",
       "26   saturn_mini_borg_like  SR-VP_05_06_2024_coassembly_Saturn_mini-Borg-l...\n",
       "27       ruby_borg_related  SR-VP_05_06_2024_coassembly_Ruby-Borg-related_...\n",
       "28              black_borg      SR-VP_05_06_2024_coassembly_Black_Borg_32_272\n",
       "29         linear_ece_19kb  Final_SR-VP_05_06_2024_coassembly_19kb_linear_..."
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_id_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb04e12a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
