{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "81c7a44a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import re \n",
    "from utils import * \n",
    "import itertools \n",
    "from tqdm import tqdm \n",
    "import warnings \n",
    "import subprocess\n",
    "import os \n",
    "from src.ggkbase import * \n",
    "import glob\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1c11f465",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_genome_id_map = dict()\n",
    "new_genome_id_map['SR-VP_05_06_2024_N_bottom.metabat2.2860'] = 'SR-VP_05_06_2024_N_bottom_DPANN-maybe_43_7'\n",
    "new_genome_id_map['SR-VP_05_06_2024_N_bottom.metabat2.1582'] = 'SR-VP_05_06_2024_N_bottom_Euryarchaeota_66_35'\n",
    "new_genome_id_map['SR-VP_05_06_2024_N_bottom.metabat2.223'] = 'SR-VP_05_06_2024_N_bottom_DPANN_Archaea_34_18'\n",
    "new_genome_id_map['SR-VP_05_06_2024_N_top.vamb.17615'] = 'SR-VP_05_06_2024_N_top_Candidatus_Methanoperedens_44_7'\n",
    "new_genome_id_map['SR-VP_05_06_2024_N_top.metabat2.305'] = 'SR-VP_05_06_2024_N_top_Euryarchaeota_54_20'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e3e2093f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. Archaeal MAGs: 71\n"
     ]
    }
   ],
   "source": [
    "metadata_df = load_metadata()\n",
    "metadata_df = metadata_df[metadata_df.taxonomy.str.contains('Archaea')].copy() # Include only the archaea. \n",
    "print('Num. Archaeal MAGs:', len(np.unique(metadata_df.index)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a00cc500",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrieving specified genomes from the contig file...: 100%|██████████| 28/28 [00:23<00:00,  1.17it/s]\n"
     ]
    }
   ],
   "source": [
    "sample = 'top'\n",
    "\n",
    "contigs_path = f'../data/contigs/n_{sample}_contigs.fa'\n",
    "scaffold_to_bin_path = f'../data/n_{sample}_scaffold_to_bin.tsv'\n",
    "\n",
    "metadata_df = load_metadata()\n",
    "metadata_df = metadata_df[(metadata_df.index.str.contains(f'N_{sample}'))].copy()\n",
    "metadata_df = metadata_df[metadata_df.taxonomy.str.contains('Archaea')].copy() # Include only the archaea. \n",
    "metadata_df.index = [new_genome_id_map.get(genome_id, genome_id) for genome_id in metadata_df.index]\n",
    "\n",
    "for genome_id in tqdm(np.unique(metadata_df.index), desc='Retrieving specified genomes from the contig file...'):\n",
    "    contigs_path_ = contigs_remove_unknown(path=contigs_path)\n",
    "    scaffold_to_bin_path_ = scaffold_to_bin_remove_unknown(path=scaffold_to_bin_path)\n",
    "    contigs_get_genome(genome_id, contigs_path=contigs_path_, scaffold_to_bin_path=scaffold_to_bin_path_)\n",
    "\n",
    "# genome_id = 'SR-VP_05_06_2024_N_bottom.metabat2.4162'\n",
    "# get_scaffold_ids(genome_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e9408475",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "run_prodigal:   0%|          | 0/48 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "run_prodigal: 100%|██████████| 48/48 [00:00<00:00, 3380.40it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run_prodigal: Could not run Prodigal on ../data/genomes/SR-VP_05_06_2024_N_top_maxbin2_40.fn\n",
      "run_prodigal: Could not run Prodigal on ../data/genomes/SR-VP_05_06_2024_N_top_maxbin2_12.fn\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def run_prodigal(fn_dir:str='../data/genomes/', fa_dir='../data/proteins/'):\n",
    "    for path in tqdm(glob.glob(os.path.join(fn_dir, '*')), desc='run_prodigal'):\n",
    "        fa_path = re.sub('.fn|.fna', '.fa', os.path.basename(path))\n",
    "        fa_path = os.path.join(fa_dir, fa_path)\n",
    "        try:\n",
    "            if not os.path.exists(fa_path):\n",
    "                subprocess.run(f'prodigal -i {path} -a {fa_path}', shell=True, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "        except:\n",
    "            print(f'run_prodigal: Could not run Prodigal on {path}')\n",
    "\n",
    "run_prodigal()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a42ec349",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
