{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "714464d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "from utils import * \n",
    "import tarfile\n",
    "import json\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a34d54bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "linear_ece_19kb_df = FASTAFile.from_file('../data/data/linear_ece_19kb.fa').to_df(parse_description=False)\n",
    "linear_ece_19kb_df.index = [id_.split('.')[-1] for id_ in linear_ece_19kb_df.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "89c822fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def foldseek_load(path):\n",
    "   '''This loads the result of running Foldseek against all databases accessible on the Foldseek web interface.'''\n",
    "   archive = tarfile.open(path, mode='r:gz')\n",
    "   cols = ['id_','subject_id','identity','alignment_length', 'mismatch', 'gap_open', 'query_alignment_start', 'query_alignment_stop', 'subject_alignment_start', 'subject_alignment_stop']\n",
    "   cols += ['tm_score', 'e_value','bit_score', 'query_length', 'subject_length',  'query_alignment', 'subject_alignment', 'per_residue_lddt', 'subject_seq', 'subject_taxonomy_id', 'subject_species_name']\n",
    "   \n",
    "   foldseek_df = list()\n",
    "   for member in archive.getmembers():\n",
    "      if ('report' in member.name):\n",
    "         continue \n",
    "      f = archive.extractfile(member)\n",
    "      content = f.read().decode('utf-8')\n",
    "      df = pd.read_csv(io.StringIO(content), sep='\\t', names=cols)\n",
    "      df['database'] = member.name.replace('.m8', '').replace('alis_', '')\n",
    "      if len(df) > 0:\n",
    "         foldseek_df.append(df)\n",
    "\n",
    "   if len(foldseek_df) == 0:\n",
    "      return pd.DataFrame([])\n",
    "   \n",
    "   foldseek_df = pd.concat(foldseek_df).reset_index(drop=True)\n",
    "   foldseek_df['id_'] = re.search(r'\\d+_\\d+', path).group(0)\n",
    "   foldseek_df['subject_species_name'] = foldseek_df['subject_species_name'].fillna('none')\n",
    "   return foldseek_df\n",
    "\n",
    "foldseek_df = pd.concat([foldseek_load(path) for path in glob.glob('../data/foldseek/esmfold/*')])\n",
    "foldseek_df = pd.concat([foldseek_df] + [foldseek_load(path) for path in glob.glob('../data/foldseek/alphafold/*')])\n",
    "# foldseek_df.to_csv('../data/foldseek/linear_ece_19kb.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "1ee24d22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. linear_ece_19kb proteins with hits: 30\n",
      "Num. linear_ece_19kb proteins with hits passing the filter: 21\n"
     ]
    }
   ],
   "source": [
    "foldseek_df['query_original_length'] = foldseek_df.id_.map(linear_ece_19kb_df.seq.apply(len)) # Foldseek search was done on trimmed structures, so want to re-add original sequence length. \n",
    "foldseek_df['query_alignment_length'] = (foldseek_df.query_alignment_stop - foldseek_df.query_alignment_start)\n",
    "foldseek_df['query_coverage'] = foldseek_df.query_alignment_length / foldseek_df.query_original_length\n",
    "# foldseek_df = foldseek_df[~foldseek_df.query_original_length.isnull()].copy() # Some of the reverse-strand ORFs are included.\n",
    "\n",
    "print(f'Num. {ece_id} proteins with hits:', foldseek_df.id_.nunique())\n",
    "\n",
    "mask = (foldseek_df.query_alignment_length > 15) | (foldseek_df.query_coverage > 0.5)\n",
    "mask = mask & (foldseek_df.tm_score > 0.7)\n",
    "foldseek_df = foldseek_df[mask].copy()\n",
    "\n",
    "print(f'Num. {ece_id} proteins with hits passing the filter:', foldseek_df.id_.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "3ff3a5c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matched 38324 out of 38411 Foldseek hits.\n"
     ]
    }
   ],
   "source": [
    "# From ChatGPT (so take with a grain of salt), not all Mgnify (and other) IDs are in UniParc; some are directly from metagenome FASTA files. \n",
    "uniparc_id_pattern = r'(?P<id>[a-zA-Z\\d]{4,})'\n",
    "uniparc_id_pattern_protvar = r'(?P<id_1>[a-zA-Z\\d]+)(\\-[0-9])*_(?P<id_2>[a-zA-Z\\d]+)(\\-[0-9])*_(?P<chain>[A-Z])'\n",
    "uniparc_version_pattern = r'(?P<version>\\d+)'\n",
    "description_pattern = r'(?P<description>.*)'\n",
    "pdb_id_pattern = r'(?P<id>[a-z0-9]{4})'\n",
    "pdb_chain_pattern = r'(?P<chain>([A-Z0-9a-z])|([A-Z0-9a-z]-\\d))'\n",
    "pdb_assembly_pattern = r'(?P<assembly>\\d)'\n",
    "gmgc_id_pattern = r'(?P<id>GMGC.+)\\.pdb*' # These are not often not associated with UniParc (or other database) entries.\n",
    "mgnify_id_pattern = r'(?P<id>MGYP\\d+)'\n",
    "\n",
    "\n",
    "subject_id_patterns = dict() \n",
    "subject_id_patterns[rf'LevyLab_{uniparc_id_pattern}_V\\d_\\d_relaxed_{pdb_chain_pattern}'] = 'UniParc'\n",
    "subject_id_patterns[rf'AF-{uniparc_id_pattern}-{uniparc_version_pattern}-F1-model_v\\d {description_pattern}'] = 'UniParc'\n",
    "subject_id_patterns[rf'AF-{uniparc_id_pattern}-F1-model_v\\d {description_pattern}'] = 'UniParc'\n",
    "subject_id_patterns[rf'{mgnify_id_pattern}\\.pdb\\.gz'] = 'Mgnify'\n",
    "subject_id_patterns[rf'{uniparc_id_pattern}_unrelaxed_rank_\\d+_alphafold\\d_ptm_model_\\d_seed_\\d+'] = 'UniParc'\n",
    "subject_id_patterns[rf'{uniparc_id_pattern}_{uniparc_version_pattern}_unrelaxed_rank_\\d+_alphafold\\d_ptm_model_\\d_seed_\\d+'] = 'UniParc'\n",
    "subject_id_patterns[rf'af_{uniparc_id_pattern}.+'] = 'UniParc'\n",
    "subject_id_patterns[rf'{pdb_id_pattern}{pdb_chain_pattern}\\d\\d'] = 'PDB'\n",
    "subject_id_patterns[rf'{pdb_id_pattern}-assembly{pdb_assembly_pattern}\\.cif\\.gz_{pdb_chain_pattern} {description_pattern}'] = 'PDB'\n",
    "subject_id_patterns[rf'{pdb_id_pattern}-assembly{pdb_assembly_pattern}\\.cif\\.gz_{pdb_chain_pattern} {description_pattern}'] = 'PDB'\n",
    "subject_id_patterns[rf'ProtVar_{uniparc_id_pattern_protvar}'] = 'UniParc'\n",
    "subject_id_patterns[mgnify_id_pattern] = 'Mgnify'\n",
    "subject_id_patterns[gmgc_id_pattern] = 'GMGC'\n",
    "\n",
    "subject_id_df = list()\n",
    "\n",
    "def get_uniparc_id_from_protvar_id(info:dict):\n",
    "    '''In these cases, chain is not the actual chain, but seems to indicate the specific sequence in the pair.'''\n",
    "    if 'id_1' not in info:\n",
    "        return info\n",
    "    return {'id':info['id_1']} if (info['chain'] == 'A') else {'id':info['id_2']}\n",
    "\n",
    "subject_ids_no_match = list()\n",
    "\n",
    "for subject_id in foldseek_df.subject_id:\n",
    "    \n",
    "    if subject_id.startswith('ModelArchive') or subject_id.startswith('Predictome'):\n",
    "        continue # These IDs are messing things up, easier to just skip them. \n",
    "\n",
    "    for pattern, database in subject_id_patterns.items():\n",
    "        match = re.search(pattern, subject_id)\n",
    "        if match is not None:\n",
    "            info = get_uniparc_id_from_protvar_id(match.groupdict())\n",
    "            info['foldseek_id'] = subject_id # Store the original ID. \n",
    "            info['database'] = database\n",
    "            subject_id_df.append(info)\n",
    "            break\n",
    "\n",
    "subject_id_df = pd.DataFrame(subject_id_df)\n",
    "subject_id_df['id'] = [f'{row.id}_{row.chain}' if (row.database == 'PDB') else row.id for row in subject_id_df.itertuples()]\n",
    "\n",
    "print(f'Matched {len(subject_id_df)} out of {len(foldseek_df)} Foldseek hits.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "61e283a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _uniparc_get_metadata(id_:str):\n",
    "    df = list()\n",
    "\n",
    "    url = f'https://rest.uniprot.org/uniparc/search?query=\"{id_}\"'\n",
    "    result = requests.get(url).text \n",
    "    result = json.loads(result)['results'][0]\n",
    "\n",
    "    info = dict()\n",
    "    info['id'] = id_\n",
    "    info['uniparc_id'] = result['uniParcId'] # The actual U* UniParc ID. \n",
    "    # Note that this only grabs info for the first taxon.\n",
    "    info['taxonomy_toplevel'] = result['commonTaxons'][0]['topLevel']\n",
    "    info['taxonomy'] = result['commonTaxons'][0]['commonTaxon']\n",
    "    info['seq'] = result['sequence']['value']\n",
    "\n",
    "    annotations = result.get('sequenceFeatures', [])\n",
    "    info['n_annotations'] = len(annotations)\n",
    "    if len(annotations) == 0:\n",
    "        return pd.DataFrame([info])\n",
    "    \n",
    "    for annotation in annotations:\n",
    "        annotation = annotation\n",
    "        info_ = info.copy()\n",
    "        info_['annotation_interpro_id'] = annotation.get('interproGroup', {}).get('id', 'none')\n",
    "        info_['annotation_interpro_description'] = annotation.get('interproGroup', {}).get('name', 'none')\n",
    "        info_['annotation_database'] = annotation['database']\n",
    "        info_['annotation_id'] = annotation['databaseId']\n",
    "        info_['annotation_start'] = annotation['locations'][0]['start']\n",
    "        info_['annotation_stop'] = annotation['locations'][0]['end']\n",
    "        df.append(info_)\n",
    "\n",
    "    return pd.DataFrame(df)\n",
    "\n",
    "\n",
    "def uniparc_get_metadata(ids:list):\n",
    "    ids = np.unique(ids) # Make sure not to download stuff twice. \n",
    "    metadata_df = list()\n",
    "    for id_ in tqdm(ids, desc='uniparc_get_metadata: Downloading UniParc metadata.'):\n",
    "        try:\n",
    "            metadata_df.append(_uniparc_get_metadata(id_))\n",
    "        except:\n",
    "            print(f'uniparc_get_metadata: Failed on ID {id_}.')\n",
    "    metadata_df = pd.concat(metadata_df)\n",
    "    return metadata_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "b5c89c7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 140 PDB entries with no UniProt accession.\n"
     ]
    }
   ],
   "source": [
    "def _pdb_get_mappings(id_:str):\n",
    "    id_, chain = id_.split('_')\n",
    "    url = f'https://www.ebi.ac.uk/pdbe/api/mappings/uniprot/{id_}'\n",
    "    print(requests.get(url).text)\n",
    "    result = json.loads(requests.get(url).text)[id_]['UniProt']\n",
    "    result = {uniprot_id:mappings['mappings'] for uniprot_id, mappings in result.items()}\n",
    "    mappings = {f\"{id_}_{mapping['chain_id']}\":uniprot_id  for uniprot_id, mappings in result.items() for mapping in mappings}\n",
    "    return mappings\n",
    "\n",
    "\n",
    "def pdb_get_mappings(ids:pd.DataFrame, path:str='../data/foldseek/pdb-uniprot_map.json'):\n",
    "    if os.path.exists(path):\n",
    "         with open(path, 'rb') as f:\n",
    "            mappings = json.load(f)\n",
    "            return mappings\n",
    "         \n",
    "    mappings = dict()\n",
    "    for id_ in tqdm(ids, desc='pdb_get_mappings: Downloading PDB mappings.'):\n",
    "        try:\n",
    "            mappings.update(_pdb_get_mappings(id_))\n",
    "        except:\n",
    "            print(f'pdb_get_mappings: Failed on PDB ID {id_}.')\n",
    "    with open(path, 'w') as f:\n",
    "            json.dump(mappings, f)\n",
    "    return mappings\n",
    "\n",
    "ids = subject_id_df[subject_id_df.database == 'PDB']['id'].unique() \n",
    "pdb_mappings = pdb_get_mappings(ids)\n",
    "\n",
    "subject_id_df['pdb_id'] = np.where(subject_id_df.database == 'PDB', subject_id_df['id'], 'none')\n",
    "subject_id_df['id'] = np.where(subject_id_df.database == 'PDB', subject_id_df['id'].map(pdb_mappings), subject_id_df['id'])\n",
    "print(f'Removing {subject_id_df['id'].isnull().sum()} PDB entries with no UniProt accession.')\n",
    "subject_id_df = subject_id_df[~subject_id_df['id'].isnull()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "2e2ba24d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('../data/foldseek/uniparc_metadata.csv'):\n",
    "    ids = subject_id_df[subject_id_df.database.isin(['UniParc', 'PDB'])]['id'].tolist()\n",
    "    uniparc_metadata_df = uniparc_get_metadata(ids)\n",
    "    uniparc_metadata_df.to_csv('../data/foldseek/uniparc_metadata.csv')\n",
    "uniparc_metadata_df = pd.read_csv('../data/foldseek/uniparc_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "61ae51cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>id</th>\n",
       "      <th>uniparc_id</th>\n",
       "      <th>taxonomy_toplevel</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>seq</th>\n",
       "      <th>n_annotations</th>\n",
       "      <th>annotation_interpro_id</th>\n",
       "      <th>annotation_interpro_description</th>\n",
       "      <th>annotation_database</th>\n",
       "      <th>annotation_id</th>\n",
       "      <th>annotation_start</th>\n",
       "      <th>annotation_stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A009IFH5</td>\n",
       "      <td>UPI0004484EB2</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...</td>\n",
       "      <td>3</td>\n",
       "      <td>IPR027417</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "      <td>SUPFAM</td>\n",
       "      <td>SSF52540</td>\n",
       "      <td>6.0</td>\n",
       "      <td>117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A0A009IFH5</td>\n",
       "      <td>UPI0004484EB2</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...</td>\n",
       "      <td>3</td>\n",
       "      <td>IPR027417</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "      <td>Gene3D</td>\n",
       "      <td>G3DSA:3.40.50.300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A0A009IFH5</td>\n",
       "      <td>UPI0004484EB2</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...</td>\n",
       "      <td>3</td>\n",
       "      <td>IPR008900</td>\n",
       "      <td>Zona occludens toxin, N-terminal</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF05707</td>\n",
       "      <td>43.0</td>\n",
       "      <td>175.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A009IPH3</td>\n",
       "      <td>UPI00044793A4</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Acinetobacter calcoaceticus/baumannii complex</td>\n",
       "      <td>MIYLITATPGSGKTLWAVKEIFARANEAEPWNIFSNIDGLKLDTAQ...</td>\n",
       "      <td>2</td>\n",
       "      <td>IPR027417</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "      <td>Gene3D</td>\n",
       "      <td>G3DSA:3.40.50.300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A0A009IPH3</td>\n",
       "      <td>UPI00044793A4</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Acinetobacter calcoaceticus/baumannii complex</td>\n",
       "      <td>MIYLITATPGSGKTLWAVKEIFARANEAEPWNIFSNIDGLKLDTAQ...</td>\n",
       "      <td>2</td>\n",
       "      <td>IPR008900</td>\n",
       "      <td>Zona occludens toxin, N-terminal</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF05707</td>\n",
       "      <td>59.0</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97891</th>\n",
       "      <td>0</td>\n",
       "      <td>Z9JXA4</td>\n",
       "      <td>UPI00044C5CEB</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Brachybacterium phenoliresistens</td>\n",
       "      <td>MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...</td>\n",
       "      <td>5</td>\n",
       "      <td>IPR027417</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "      <td>Gene3D</td>\n",
       "      <td>G3DSA:3.40.50.300</td>\n",
       "      <td>121.0</td>\n",
       "      <td>327.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97892</th>\n",
       "      <td>1</td>\n",
       "      <td>Z9JXA4</td>\n",
       "      <td>UPI00044C5CEB</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Brachybacterium phenoliresistens</td>\n",
       "      <td>MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...</td>\n",
       "      <td>5</td>\n",
       "      <td>IPR003714</td>\n",
       "      <td>PhoH-like protein</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02562</td>\n",
       "      <td>124.0</td>\n",
       "      <td>326.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97893</th>\n",
       "      <td>2</td>\n",
       "      <td>Z9JXA4</td>\n",
       "      <td>UPI00044C5CEB</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Brachybacterium phenoliresistens</td>\n",
       "      <td>MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...</td>\n",
       "      <td>5</td>\n",
       "      <td>IPR027417</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "      <td>SUPFAM</td>\n",
       "      <td>SSF52540</td>\n",
       "      <td>123.0</td>\n",
       "      <td>290.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97894</th>\n",
       "      <td>3</td>\n",
       "      <td>Z9JXA4</td>\n",
       "      <td>UPI00044C5CEB</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Brachybacterium phenoliresistens</td>\n",
       "      <td>MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...</td>\n",
       "      <td>5</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.300:FF:000013</td>\n",
       "      <td>121.0</td>\n",
       "      <td>325.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97895</th>\n",
       "      <td>4</td>\n",
       "      <td>Z9JXA4</td>\n",
       "      <td>UPI00044C5CEB</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>Brachybacterium phenoliresistens</td>\n",
       "      <td>MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...</td>\n",
       "      <td>5</td>\n",
       "      <td>IPR051451</td>\n",
       "      <td>PhoH2-like</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR30473</td>\n",
       "      <td>16.0</td>\n",
       "      <td>329.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97896 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0          id     uniparc_id   taxonomy_toplevel  \\\n",
       "0               0  A0A009IFH5  UPI0004484EB2  cellular organisms   \n",
       "1               1  A0A009IFH5  UPI0004484EB2  cellular organisms   \n",
       "2               2  A0A009IFH5  UPI0004484EB2  cellular organisms   \n",
       "3               0  A0A009IPH3  UPI00044793A4  cellular organisms   \n",
       "4               1  A0A009IPH3  UPI00044793A4  cellular organisms   \n",
       "...           ...         ...            ...                 ...   \n",
       "97891           0      Z9JXA4  UPI00044C5CEB  cellular organisms   \n",
       "97892           1      Z9JXA4  UPI00044C5CEB  cellular organisms   \n",
       "97893           2      Z9JXA4  UPI00044C5CEB  cellular organisms   \n",
       "97894           3      Z9JXA4  UPI00044C5CEB  cellular organisms   \n",
       "97895           4      Z9JXA4  UPI00044C5CEB  cellular organisms   \n",
       "\n",
       "                                            taxonomy  \\\n",
       "0                            Acinetobacter baumannii   \n",
       "1                            Acinetobacter baumannii   \n",
       "2                            Acinetobacter baumannii   \n",
       "3      Acinetobacter calcoaceticus/baumannii complex   \n",
       "4      Acinetobacter calcoaceticus/baumannii complex   \n",
       "...                                              ...   \n",
       "97891               Brachybacterium phenoliresistens   \n",
       "97892               Brachybacterium phenoliresistens   \n",
       "97893               Brachybacterium phenoliresistens   \n",
       "97894               Brachybacterium phenoliresistens   \n",
       "97895               Brachybacterium phenoliresistens   \n",
       "\n",
       "                                                     seq  n_annotations  \\\n",
       "0      MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...              3   \n",
       "1      MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...              3   \n",
       "2      MARIRLTTGGIGAGKTYLNVKLADEAHKKGQYTKIYSNIRAHSELT...              3   \n",
       "3      MIYLITATPGSGKTLWAVKEIFARANEAEPWNIFSNIDGLKLDTAQ...              2   \n",
       "4      MIYLITATPGSGKTLWAVKEIFARANEAEPWNIFSNIDGLKLDTAQ...              2   \n",
       "...                                                  ...            ...   \n",
       "97891  MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...              5   \n",
       "97892  MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...              5   \n",
       "97893  MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...              5   \n",
       "97894  MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...              5   \n",
       "97895  MTDLESVPGRAPEVRERRLAIPDHLSPFQVLGENDQALAALEDVVP...              5   \n",
       "\n",
       "      annotation_interpro_id  \\\n",
       "0                  IPR027417   \n",
       "1                  IPR027417   \n",
       "2                  IPR008900   \n",
       "3                  IPR027417   \n",
       "4                  IPR008900   \n",
       "...                      ...   \n",
       "97891              IPR027417   \n",
       "97892              IPR003714   \n",
       "97893              IPR027417   \n",
       "97894                   none   \n",
       "97895              IPR051451   \n",
       "\n",
       "                         annotation_interpro_description annotation_database  \\\n",
       "0      P-loop containing nucleoside triphosphate hydr...              SUPFAM   \n",
       "1      P-loop containing nucleoside triphosphate hydr...              Gene3D   \n",
       "2                       Zona occludens toxin, N-terminal                Pfam   \n",
       "3      P-loop containing nucleoside triphosphate hydr...              Gene3D   \n",
       "4                       Zona occludens toxin, N-terminal                Pfam   \n",
       "...                                                  ...                 ...   \n",
       "97891  P-loop containing nucleoside triphosphate hydr...              Gene3D   \n",
       "97892                                  PhoH-like protein                Pfam   \n",
       "97893  P-loop containing nucleoside triphosphate hydr...              SUPFAM   \n",
       "97894                                               none              FunFam   \n",
       "97895                                         PhoH2-like             PANTHER   \n",
       "\n",
       "                     annotation_id  annotation_start  annotation_stop  \n",
       "0                         SSF52540               6.0            117.0  \n",
       "1                G3DSA:3.40.50.300               1.0            174.0  \n",
       "2                          PF05707              43.0            175.0  \n",
       "3                G3DSA:3.40.50.300               1.0            174.0  \n",
       "4                          PF05707              59.0            172.0  \n",
       "...                            ...               ...              ...  \n",
       "97891            G3DSA:3.40.50.300             121.0            327.0  \n",
       "97892                      PF02562             124.0            326.0  \n",
       "97893                     SSF52540             123.0            290.0  \n",
       "97894  G3DSA:3.40.50.300:FF:000013             121.0            325.0  \n",
       "97895                    PTHR30473              16.0            329.0  \n",
       "\n",
       "[97896 rows x 13 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniparc_metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "682b5843",
   "metadata": {},
   "outputs": [],
   "source": [
    "foldseek_df['id'] = foldseek_df['subject_id'].map(subject_id_df.drop_duplicates('foldseek_id').set_index('foldseek_id')['id'])\n",
    "foldseek_df['subject_database'] = foldseek_df['id'].map(subject_id_df.drop_duplicates('id').set_index('id')['database'])\n",
    "foldseek_df['subject_description'] = foldseek_df['id'].map(subject_id_df.drop_duplicates('id').set_index('id')['description'])\n",
    "\n",
    "if 'annotation_database' not in foldseek_df.columns:\n",
    "    foldseek_df = foldseek_df.merge(uniparc_metadata_df, left_on='id', right_on='id', how='left')\n",
    "\n",
    "get_annotation_overlap = lambda row : max(min(row.subject_alignment_stop, row.annotation_stop) - max(row.subject_alignment_start, row.annotation_start), 0)\n",
    "has_annotation = lambda row : row.annotation_start in np.arange(0, 1000)\n",
    "foldseek_df['annotation_overlap'] = [get_annotation_overlap(row) if has_annotation(row) else np.nan for row in foldseek_df.itertuples()]\n",
    "foldseek_df['annotation_coverage'] = foldseek_df.annotation_overlap / (foldseek_df.annotation_stop - foldseek_df.annotation_start)\n",
    "foldseek_df['query_coverage'] = (foldseek_df.query_alignment_stop - foldseek_df.query_alignment_start) / foldseek_df.query_length\n",
    "\n",
    "# Remove annotations that don't overlap the alignment. \n",
    "# mask = (foldseek_df.n_annotations > 1) & (foldseek_df.annotation_overlap < 10)\n",
    "# print(f'Removing {mask.sum()} useless annotations.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "1189cab7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ECE genes with no good hits: 1_4 1_5 1_6 1_7 1_9 1_13 1_15 1_16 1_18 1_22 1_24 1_27 1_29 1_30 1_31 1_34 1_36\n"
     ]
    }
   ],
   "source": [
    "ece_gene_ids = [f'1_{i + 1}' for i in range(38)]\n",
    "print('ECE genes with no good hits:', ' '.join([gene_id for gene_id in ece_gene_ids if (gene_id not in foldseek_df.id_.unique())]))\n",
    "ece_gene_ids_no_hits = '1_4 1_5 1_6 1_7 1_9 1_13 1_15 1_16 1_18 1_22 1_24 1_27 1_29 1_30 1_31 1_34 1_36'.split()\n",
    "# len(ece_gene_ids_no_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41225566",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query sequence length: 44 (trimmed length 25)\n",
      "Num hits: 107\n",
      "Num. Mgnify hits: 27\n",
      "Num. hits to uncharacterized proteins: 30\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "annotation_interpro_description\n",
       "Zinc finger C2H2-type                                            141\n",
       "none                                                             103\n",
       "Zinc finger C2H2 superfamily                                      28\n",
       "ROS/MUCR transcriptional regulator superfamily                    11\n",
       "Zinc Finger C2H2-type Transcription Regulators                    10\n",
       "Zinc finger, AD-type                                               8\n",
       "ROS/MUCR transcriptional regulator                                 6\n",
       "Drought induced 19 protein type, zinc-binding domain               5\n",
       "SANT/Myb domain                                                    4\n",
       "BTB/POZ domain                                                     3\n",
       "Zinc finger                                                        3\n",
       "Snail/Krueppel C2H2-type Zinc-finger                               3\n",
       "SKP1/BTB/POZ domain superfamily                                    2\n",
       "Insect cuticle protein                                             2\n",
       "Rubredoxin, iron-binding site                                      2\n",
       "Homedomain-like superfamily                                        2\n",
       "Zinc finger transcriptional regulators                             2\n",
       "SET domain superfamily                                             1\n",
       "Chitin-binding type R&R consensus                                  1\n",
       "Regulatory Factors in Gene Expression and Cell Death               1\n",
       "Krueppel C2H2-type Zinc Finger                                     1\n",
       "Zinc finger and ubiquitin peptidase domain-containing protein      1\n",
       "Zinc finger and BTB domain-containing                              1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def print_info(id_:str, min_annotation_overlap:int=10, min_annotation_coverage:float=0.5, annotations_only:bool=False):\n",
    "    df = foldseek_df[foldseek_df.id_ == id_].copy()\n",
    "    print('Query sequence length:', df.query_original_length.iloc[0], f'(trimmed length {df.query_length.iloc[0]})')\n",
    "    print('Num hits:', df['id'].nunique())\n",
    "    print('Num. Mgnify hits:', (df.drop_duplicates('id').subject_database == 'Mgnify').sum())\n",
    "    df = df[df.subject_database != 'Mgnify'].copy()\n",
    "    print('Num. hits to uncharacterized proteins:', (df.drop_duplicates('id').n_annotations == 0).sum())\n",
    "    if annotations_only:\n",
    "        df = df[(df.n_annotations > 0) & (df.annotation_overlap > min_annotation_overlap) & (df.annotation_coverage > min_annotation_coverage)].copy()\n",
    "    return df.sort_values('query_coverage', ascending=False)\n",
    "\n",
    "print_info('1_14')# .annotation_interpro_description.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a03e30da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id_', 'subject_id', 'identity', 'alignment_length', 'mismatch',\n",
       "       'gap_open', 'query_alignment_start', 'query_alignment_stop',\n",
       "       'subject_alignment_start', 'subject_alignment_stop', 'tm_score',\n",
       "       'e_value', 'bit_score', 'query_length', 'subject_length',\n",
       "       'query_alignment', 'subject_alignment', 'per_residue_lddt',\n",
       "       'subject_seq', 'subject_taxonomy_id', 'subject_species_name',\n",
       "       'database', 'query_original_length', 'query_alignment_length',\n",
       "       'query_coverage', 'id', 'subject_database', 'subject_description',\n",
       "       'Unnamed: 0', 'uniparc_id', 'taxonomy_toplevel', 'taxonomy', 'seq',\n",
       "       'n_annotations', 'annotation_interpro_id',\n",
       "       'annotation_interpro_description', 'annotation_database',\n",
       "       'annotation_id', 'annotation_start', 'annotation_stop',\n",
       "       'annotation_overlap', 'annotation_coverage'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foldseek_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e5bef3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
