{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bf1723ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import * \n",
    "import os \n",
    "from src.ggkbase import * \n",
    "import glob\n",
    "import matplotlib as mpl \n",
    "from src.files import BLASTFileJSON, FASTAFile, InterProScanFileTSV, GenBankFile, fasta_get_genome_size\n",
    "import subprocess\n",
    "import networkx as nx\n",
    "import json\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Notebook for all the setup stuff. \n",
    "\n",
    "# SR-VP_05_06_2024_ck_bottom_Methanoperedens_41_16 is incomplete but important, needs to be binned in other samples. \n",
    "# SR-VP_05_06_2024_ck_bottom_scaffold_71 is a possible misbinned ECE, so don't use as a seed sequence. \n",
    "# Strain seems like it might be most abundant in n_middle or ck_bottom 2025.\n",
    "\n",
    "# Mysterious McrA is on this scaffold https://ggkbase.berkeley.edu/organisms/669853/contigs/1100394534. A bit worried that it didn't end up in any of the bins.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa1a02d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mp_x_contig_ids = subprocess.run('cat ../data/mp_x.fn | grep \">\"', capture_output=True, shell=True, check=True).stdout.decode().split('\\n')\n",
    "# mp_x_contig_ids = [contig_id.replace('>', '').split()[0] for contig_id in mp_x_contig_ids if (len(contig_id) > 0)]\n",
    "# mp_x_contig_ids = [contig_id for contig_id in mp_x_contig_ids if (contig_id not in blast_df.qseqid.unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "261ecfe9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.436289)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "fasta_df = FASTAFile.from_file('../data/checkm/mp_x_n_top_2025.fn').to_df()\n",
    "fasta_df.seq.apply(len).sum() / 1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b88a8a3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. Methanoperedens genomes meeting the minimum size requirement: 94\n"
     ]
    }
   ],
   "source": [
    "# Dowloaded pretty much every Methanoperedens genome under the SR-VP project.\n",
    "mp_genome_paths = glob.glob('../data/ggkbase/contigs/methanoperedens/*')\n",
    "mp_genome_ids = [os.path.basename(path).replace('.contigs.fa', '') for path in mp_genome_paths]\n",
    "mp_genome_sizes = [fasta_get_genome_size(path) for path in mp_genome_paths]\n",
    "\n",
    "mp_genome_paths = np.array(mp_genome_paths)[np.array(mp_genome_sizes) > 1e6].tolist()\n",
    "print('Num. Methanoperedens genomes meeting the minimum size requirement:', len(mp_genome_paths))\n",
    "\n",
    "for path in glob.glob('../data/ggkbase/contigs/methanoperedens/*'):\n",
    "    if path not in mp_genome_paths:\n",
    "        os.remove(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af01a67d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num. Methanoperedens strains: 20\n"
     ]
    }
   ],
   "source": [
    "# Genomes selected after running dRep with default parameters. \n",
    "mp_genome_ids = [os.path.basename(path).replace('.contigs.fa', '') for path in glob.glob('../data/ggkbase/drep/dereplicated_genomes/*')]\n",
    "print('Num. Methanoperedens strains:', len(mp_genome_ids))\n",
    "\n",
    "genome_id_to_ggkbase_id_map = {f'mp_{i + 1}':genome_id for i, genome_id in enumerate(mp_genome_ids)}\n",
    "\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_1'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_33_21'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_2'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_32_5'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_3'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_34_1246'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_4'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_36_6'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_6'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_35_3'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_7'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_33_6'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_8'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_31_4'\n",
    "genome_id_to_ggkbase_id_map['jupiter_mini_borg_9'] = 'SR-VP_05_06_2024_coassembly_Jupiter_mini-Borg_35_6'\n",
    "genome_id_to_ggkbase_id_map['saturn_mini_borg_1'] = 'SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_35_7'\n",
    "genome_id_to_ggkbase_id_map['saturn_mini_borg_2'] = 'SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_32_200'\n",
    "genome_id_to_ggkbase_id_map['saturn_mini_borg_3'] = 'SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_3'\n",
    "genome_id_to_ggkbase_id_map['saturn_mini_borg_4'] = 'SR-VP_05_06_2024_coassembly_Saturn_mini-Borg_33_4'\n",
    "genome_id_to_ggkbase_id_map['unclassified_mini_borg'] = 'SR-VP_05_06_2024_coassembly_mini-Borg_reminscent_42_6-8'\n",
    "genome_id_to_ggkbase_id_map['unclassified_borg'] = 'SR-VP_05_06_2024_coassembly_new_Borg_34_11'\n",
    "genome_id_to_ggkbase_id_map['amethyst_borg'] = 'SR-VP_05_06_2024_coassembly_Amethyst_Borg_34_3'\n",
    "genome_id_to_ggkbase_id_map['oxblood_borg'] = 'SR-VP_05_06_2024_coassembly_Oxblood_Borg_33_20'\n",
    "genome_id_to_ggkbase_id_map['pink_borg'] = 'SR-VP_05_06_2024_coassembly_Pink_Borg_32_55'\n",
    "genome_id_to_ggkbase_id_map['purple_borg'] = 'SR-VP_05_06_2024_coassembly_Purple_Borg_33_3'\n",
    "genome_id_to_ggkbase_id_map['rose_borg'] = 'SR-VP_05_06_2024_coassembly_Rose_Borg_31_2'\n",
    "genome_id_to_ggkbase_id_map['vermilion_borg'] = 'SR-VP_05_06_2024_coassembly_Vermilion_Borg_34_8'\n",
    "genome_id_to_ggkbase_id_map['mercury_mini_borg'] = 'SR-VP_05_06_2024_coassembly_Mercury_mini-Borg_37_9'\n",
    "genome_id_to_ggkbase_id_map['saturn_mini_borg_like'] = 'SR-VP_05_06_2024_coassembly_Saturn_mini-Borg-like_32_7'\n",
    "genome_id_to_ggkbase_id_map['ruby_borg_related'] = 'SR-VP_05_06_2024_coassembly_Ruby-Borg-related_37_10'\n",
    "genome_id_to_ggkbase_id_map['black_borg'] = 'SR-VP_05_06_2024_coassembly_Black_Borg_32_272'\n",
    "genome_id_to_ggkbase_id_map['linear_ece_19kb'] = 'Final_SR-VP_05_06_2024_coassembly_19kb_linear_ECE_26_1334_complete'\n",
    "\n",
    "genome_metadata_df = pd.DataFrame.from_dict(genome_id_to_ggkbase_id_map, orient='index').reset_index()\n",
    "genome_metadata_df.columns = ['genome_id', 'ggkbase_id']\n",
    "genome_metadata_df = genome_metadata_df.set_index('genome_id')\n",
    "genome_metadata_df.to_csv('../src/data/genome_metadata.csv')\n",
    "genome_metadata_df.to_csv('./data/genome_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ff277b48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ECE host genome: mp_18\n"
     ]
    }
   ],
   "source": [
    "# SR-VP_05_06_2024_ck_bottom_Black_host_Methanoperedens-related_44_24 \n",
    "# SR-VP_05_06_2024_N_top_Candidatus_Methanoperedens_Black-host_type_44_27 \n",
    "\n",
    "# FINAL_2-50MB_Methanoperedens_SR-VP_07_25_2022_A1_90cm_PACBIO-HIFI_44_8_complete\n",
    "\n",
    "ggkbase_id_to_genome_id_map = {ggkbase_id:genome_id for genome_id, ggkbase_id in genome_id_to_ggkbase_id_map.items()}\n",
    "print('ECE host genome:', ggkbase_id_to_genome_id_map['FINAL_2-50MB_Methanoperedens_SR-VP_07_25_2022_A1_90cm_PACBIO-HIFI_44_8_complete'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d02e894",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_contig_index = lambda gene_id : gene_id.split('.')[-1].split('_')[0]\n",
    "\n",
    "for genome_id, ggkbase_id in genome_id_to_ggkbase_id_map.items():\n",
    "    try:\n",
    "        fasta_file = FASTAFile.from_file(f'../data/ggkbase/contigs/{ggkbase_id}.contigs.fa')\n",
    "    except FileNotFoundError:\n",
    "        fasta_file = FASTAFile.from_file(f'../data/ggkbase/contigs/methanoperedens/{ggkbase_id}.contigs.fa')\n",
    "        \n",
    "    fasta_file.ids = [f'{genome_id}.{i + 1}' for i in range(len(fasta_file.ids))] # Rename the contigs\n",
    "    fasta_file.write(f'../data/data/{genome_id}.fn')\n",
    "\n",
    "    genbank_file = GenBankFile.from_file(f'../data/ggkbase/genbank/{ggkbase_id}.gbk')\n",
    "    genbank_file.df['gene_id'] = [f'{genome_id}.{gene_id}' for gene_id in genbank_file.df.gene_id] # Rename the gene IDs.\n",
    "    genbank_file.df['contig_id'] = [f'{genome_id}.{get_contig_index(gene_id)}' for gene_id in genbank_file.df.gene_id] # Rename the contig IDs.\n",
    "    subprocess.run(f'cp ../data/ggkbase/genbank/{ggkbase_id}.gbk ../data/data/{genome_id}.gbk', check=True, shell=True)\n",
    "    genbank_file.to_gff(f'../data/data/{genome_id}.gff')\n",
    "    genbank_file.to_fasta(f'../data/data/{genome_id}.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19a619b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/data/all.fn', 'w') as f:\n",
    "    paths = [path for path in glob.glob('../data/data/*fn') if (os.path.basename(path) != 'all.fn')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read())\n",
    "        \n",
    "with open('../data/data/all.fa', 'w') as f:\n",
    "    paths = [path for path in glob.glob('../data/data/*fa') if (os.path.basename(path) != 'all.fa')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read())\n",
    "\n",
    "gff_header = '##gff-version  3\\n'\n",
    "with open('../data/data/all.gff', 'w') as f:\n",
    "    f.write(gff_header)\n",
    "    paths = [path for path in glob.glob('../data/data/*gff') if (os.path.basename(path) != 'all.gff')]\n",
    "    for path in paths:\n",
    "        with open(path, 'r') as _f: \n",
    "            f.write(_f.read().replace(gff_header, ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a7d74003",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def run_fastani(input_paths:list, output_dir:str='../data/ggkbase', fragment_length:int=1000, min_fraction:float=0.2):\n",
    "#     input_list_path = os.path.join('.', 'fastani_input.txt')\n",
    "#     output_path = os.path.join(output_dir, 'fastani.tsv')\n",
    "#     with open(input_list_path, 'w') as f:\n",
    "#         f.write('\\n'.join(input_paths))\n",
    "#     cmd = f'fastANI --ql {input_list_path} --rl {input_list_path} -o {output_path} --fragLen {fragment_length} --minFraction {min_fraction}'\n",
    "#     subprocess.run(cmd, shell=True, check=True)\n",
    "#     os.remove(input_list_path)\n",
    "    \n",
    "# # run_fastani(input_paths=mp_genome_paths)\n",
    "\n",
    "# def load_fastani(path:str='../data/ggkbase/fastani.tsv'):\n",
    "#     cols = ['query_genome_path', 'ref_genome_path', 'ani', 'n_aligned_fragments', 'n_fragments']\n",
    "#     fastani_df = pd.read_csv(path, sep='\\t', names=cols)\n",
    "#     fastani_df['query_genome_id'] = fastani_df.query_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "#     fastani_df['ref_genome_id'] = fastani_df.ref_genome_path.apply(lambda path : os.path.basename(path).replace('.contigs.fa', ''))\n",
    "#     return fastani_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb04e12a",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'coverm_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[11]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mcoverm_df\u001b[49m\n",
      "\u001b[31mNameError\u001b[39m: name 'coverm_df' is not defined"
     ]
    }
   ],
   "source": [
    "coverm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c7f879",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
