{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b01ae56f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import glob \n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "import os \n",
    "from utils import * \n",
    "import subprocess\n",
    "import itertools\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "798d7707",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'sample_metadata.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[13]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m sample_metadata_df = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43msample_metadata.csv\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      2\u001b[39m sample_metadata_df[\u001b[33m'\u001b[39m\u001b[33mmetat\u001b[39m\u001b[33m'\u001b[39m] = sample_metadata_df.sample_id.str.contains(\u001b[33m'\u001b[39m\u001b[33mmetat\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m      4\u001b[39m ggkbase_id_to_year_map = sample_metadata_df.set_index(\u001b[33m'\u001b[39m\u001b[33mggkbase_id\u001b[39m\u001b[33m'\u001b[39m).year.to_dict()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/banfield/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/banfield/lib/python3.14/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    617\u001b[39m _validate_names(kwds.get(\u001b[33m\"\u001b[39m\u001b[33mnames\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/banfield/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1617\u001b[39m     \u001b[38;5;28mself\u001b[39m.options[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m] = kwds[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/banfield/lib/python3.14/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1878\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1891\u001b[39m f = \u001b[38;5;28mself\u001b[39m.handles.handle\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/banfield/lib/python3.14/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    869\u001b[39m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[32m    870\u001b[39m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[32m    871\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m         handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[32m    882\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(handle, ioargs.mode)\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: 'sample_metadata.csv'"
     ]
    }
   ],
   "source": [
    "sample_metadata_df = pd.read_csv('sample_metadata.csv')\n",
    "sample_metadata_df['metat'] = sample_metadata_df.sample_id.str.contains('metat')\n",
    "\n",
    "ggkbase_id_to_year_map = sample_metadata_df.set_index('ggkbase_id').year.to_dict()\n",
    "sample_id_to_year_map = sample_metadata_df.set_index('sample_id').year.to_dict()\n",
    "ggkbase_id_to_sample_id_map = sample_metadata_df.set_index('ggkbase_id').sample_id.to_dict()\n",
    "sample_id_to_ggkbase_id_map = sample_metadata_df.set_index('sample_id').ggkbase_id.to_dict()\n",
    "\n",
    "genome_ids = pd.read_csv('./data/genome_metadata.csv').genome_id.values\n",
    "sample_ids = sample_metadata_df.sample_id.unique()\n",
    "metat_sample_ids = sample_metadata_df[sample_metadata_df.metat].sample_id.unique()\n",
    "\n",
    "sample_path_template = '/groups/banfield/sequences/{year}/{ggkbase_id}/raw.d/{ggkbase_id}_trim_clean.{paired_end}.fastq.gz'\n",
    "ref_genome_dir = '~/data/'\n",
    "# ref_genome_paths = [os.path.join(ref_genome_dir, f'{genome_id}.fn') for genome_id in genome_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d91d6fa0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch --wrap \"coverm contig -c /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_ck_bottom/raw.d/SR-VP_05_06_2024_ck_bottom_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2025/SR-VP_05_06_2024_ck_bottom/raw.d/SR-VP_05_06_2024_ck_bottom_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.2.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.1.fastq.gz /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.2.fastq.gz -r ~/data/all.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o /home/philippar/data/coverm/all.tsv\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "coverm_sample_paths = '' \n",
    "for row in sample_metadata_df[~sample_metadata_df.metat].itertuples():\n",
    "    coverm_sample_paths += sample_path_template.format(ggkbase_id=row.ggkbase_id, paired_end='PE.1', year=row.year) + ' '\n",
    "    coverm_sample_paths += sample_path_template.format(ggkbase_id=row.ggkbase_id, paired_end='PE.2', year=row.year) + ' '\n",
    "coverm_sample_paths = coverm_sample_paths.strip()\n",
    "\n",
    "coverm_fields = 'mean trimmed_mean covered_bases variance count rpkm tpm'\n",
    "\n",
    "ref_genome_path = os.path.join(ref_genome_dir, f'all.fn')\n",
    "output_path = f'/home/philippar/data/coverm/all.tsv'\n",
    "cmd = f'coverm contig -c {coverm_sample_paths} -r {ref_genome_path} --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m {coverm_fields} -t 20 -o {output_path}'\n",
    "print(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b4da042",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../scripts/bbduk_library_sizes.sh', 'w') as f:\n",
    "#     for name, sample_id in sample_id_map.items():\n",
    "#             paired_ends_1_path = sample_path_template.format(paired_end='PE.1', sample_id=sample_id, year=sample_year_map.get(sample_id, 2025))\n",
    "#             paired_ends_2_path = sample_path_template.format(paired_end='PE.2', sample_id=sample_id, year=sample_year_map.get(sample_id, 2025))\n",
    "#             output_path = f'/home/philippar/data/bbduk/{name}.txt'\n",
    "#             cmd = f'bbduk.sh in={paired_ends_1_path} in2={paired_ends_2_path} out=/dev/null stats={output_path} threads=64'\n",
    "#             f.write(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f3c84ab3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/ck_bottom_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/ck_middle_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/ck_top_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_top_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_bottom_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025_metaT/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_middle_2025_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_ck_bot_05_06_2024_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_ck_bot_05_06_2024_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/ck_bottom_2024_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_top_05_06_2024_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_top_05_06_2024_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_top_2024_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_bot_05_06_2024_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_bot_05_06_2024_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_bottom_2024_metat.bam\n",
      "srun bbmap.sh pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam in1=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_mid_05_06_2024_metaT_trim_clean.PE.1.fastq.gz in2=/groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_06_2024_metaT/raw.d/SR-VP_Bioreactor_N_mid_05_06_2024_metaT_trim_clean.PE.2.fastq.gz ref=~/data/all.fn nodisk | shrinksam | sambam > ../data/metat/n_middle_2024_metat.bam\n"
     ]
    }
   ],
   "source": [
    "def bbmap_get_command(sample_id:str, genome_id:str, output_dir:str='../data/metat/'):\n",
    "    ref_genome_path = os.path.join(ref_genome_dir, f'{genome_id}.fn')\n",
    "\n",
    "    input_path_1 = sample_path_template.format(ggkbase_id=sample_id_to_ggkbase_id_map[sample_id], paired_end='PE.1', year=sample_id_to_year_map[sample_id])\n",
    "    input_path_2 = sample_path_template.format(ggkbase_id=sample_id_to_ggkbase_id_map[sample_id], paired_end='PE.2', year=sample_id_to_year_map[sample_id])\n",
    "\n",
    "    # output_dir = os.path.join(output_dir, genome_id)\n",
    "    output_path = os.path.join(output_dir, f'{sample_id}.bam')\n",
    "\n",
    "    params = 'pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam'\n",
    "    cmd = f'bbmap.sh {params} in1={input_path_1} in2={input_path_2} ref={ref_genome_path} nodisk | shrinksam | sambam > {output_path}'\n",
    "    return cmd, output_path\n",
    "\n",
    "\n",
    "output_dir = '../data/metat'\n",
    "\n",
    "genome_id = 'all'\n",
    "# print(f'mkdir -p {os.path.join(output_dir, genome_id)}') # Make sure the output directory exists.\n",
    "for sample_id in metat_sample_ids:        \n",
    "    cmd, output_path = bbmap_get_command(sample_id, genome_id, output_dir=output_dir)\n",
    "    # print(f'sbatch --wrap \"{cmd}\"')\n",
    "    print(f'srun {cmd}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "185f49b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/ck_bottom_2025_metat_read_counts ~/data/metat/ck_bottom_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/ck_middle_2025_metat_read_counts ~/data/metat/ck_middle_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/ck_top_2025_metat_read_counts ~/data/metat/ck_top_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_top_2025_metat_read_counts ~/data/metat/n_top_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_bottom_2025_metat_read_counts ~/data/metat/n_bottom_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_middle_2025_metat_read_counts ~/data/metat/n_middle_2025_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/ck_bottom_2024_metat_read_counts ~/data/metat/ck_bottom_2024_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_top_2024_metat_read_counts ~/data/metat/n_top_2024_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_bottom_2024_metat_read_counts ~/data/metat/n_bottom_2024_metat.bam\n",
      "featureCounts -p -T 64 -g ID -t CDS -a ~/data/all.gff -s 2 -o ~/data/metat/n_middle_2024_metat_read_counts ~/data/metat/n_middle_2024_metat.bam\n"
     ]
    }
   ],
   "source": [
    "def featurecounts_get_command(bam_path:str, ref_path:str=None, output_dir:str=None):\n",
    "    output_file_name = os.path.basename(bam_path).replace('.bam', '')\n",
    "    output_file_name += '_read_counts'\n",
    "    output_path = os.path.join(output_dir, output_file_name)\n",
    "    return f'featureCounts -p -T 64 -g ID -t CDS -a {ref_path} -s 2 -o {output_path} {bam_path}' \n",
    "\n",
    "data_dir = '~/data/metat'\n",
    "\n",
    "for sample_id in metat_sample_ids:\n",
    "    bam_path = os.path.join(data_dir, f'{sample_id}.bam')\n",
    "    ref_path = os.path.join(ref_genome_dir, 'all.gff')\n",
    "    cmd = featurecounts_get_command(bam_path, ref_path, output_dir=data_dir)\n",
    "    print(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "43c9b350",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['mp_1', 'mp_2', 'mp_3', 'mp_4', 'mp_5', 'mp_6', 'mp_7', 'mp_8',\n",
       "       'mp_9', 'mp_10', 'mp_11', 'mp_12', 'mp_13', 'mp_14', 'mp_15',\n",
       "       'mp_16', 'mp_17', 'mp_18', 'mp_19', 'mp_20', 'jupiter_mini_borg_1',\n",
       "       'jupiter_mini_borg_2', 'jupiter_mini_borg_3',\n",
       "       'jupiter_mini_borg_4', 'jupiter_mini_borg_6',\n",
       "       'jupiter_mini_borg_7', 'jupiter_mini_borg_8',\n",
       "       'jupiter_mini_borg_9', 'saturn_mini_borg_1', 'saturn_mini_borg_2',\n",
       "       'saturn_mini_borg_3', 'saturn_mini_borg_4',\n",
       "       'unclassified_mini_borg', 'unclassified_borg', 'amethyst_borg',\n",
       "       'oxblood_borg', 'pink_borg', 'purple_borg', 'rose_borg',\n",
       "       'vermilion_borg', 'mercury_mini_borg', 'saturn_mini_borg_like',\n",
       "       'ruby_borg_related', 'black_borg', 'linear_ece_19kb'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8a099aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def interproscan_get_command(input_path:str, output_dir:str='~/data/interproscan/'):\n",
    "    id_ = os.path.basename(input_path).split('.')[0]\n",
    "    output_path = os.path.join(output_dir, f'{id_}.tsv')\n",
    "    cmd = f'interproscan.sh -i {input_path} -o {output_path} -f tsv'\n",
    "    return cmd\n",
    "\n",
    "with open('../scripts/interproscan_annotation.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        input_path = f'~/data/{genome_id}.fa'\n",
    "        cmd = interproscan_get_command(input_path)\n",
    "        f.write(f'sbatch --wrap \"{cmd}\"' + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3098b1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kofamscan_get_command(input_path:str, output_dir:str='~/data/kofamscan/'):\n",
    "    id_ = os.path.basename(input_path).split('.')[0]\n",
    "    output_path = os.path.join(output_dir, f'{id_}.tsv')\n",
    "    profiles_path = '/shared/db/kegg/kofam/latest/profiles'\n",
    "    ko_list_path = '/shared/db/kegg/kofam/latest/metadata/ko_list'\n",
    "    cmd = f'/shared/software/kofamscan/latest/exec_annotation {input_path} -o {output_path} -f detail-tsv -E 0.1 --cpu 16 -k {ko_list_path} -p {profiles_path}'\n",
    "    cmd = f'sbatch --wrap \"{cmd}\"'\n",
    "    return cmd\n",
    "\n",
    "with open('../scripts/kofamscan_annotation.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        input_path = f'~/data/{genome_id}.fa'\n",
    "        cmd = kofamscan_get_command(input_path)\n",
    "        f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "672fbe9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -F 4  ~/data/metat/mp_1/ck_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/ck_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/ck_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/ck_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/ck_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/ck_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/ck_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/ck_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_top_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_top_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_1/n_middle_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_1/n_middle_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/ck_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/ck_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/ck_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/ck_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/ck_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/ck_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/ck_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/ck_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_top_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_top_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_2/n_middle_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_2/n_middle_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/ck_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/ck_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/ck_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/ck_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/ck_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/ck_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/ck_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/ck_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_top_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_top_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_3/n_middle_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_3/n_middle_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/ck_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/ck_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/ck_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/ck_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/ck_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/ck_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/ck_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/ck_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_top_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_top_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_4/n_middle_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_4/n_middle_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/ck_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/ck_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/ck_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/ck_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/ck_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/ck_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_top_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_top_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_bottom_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_bottom_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_middle_2025_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_middle_2025_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/ck_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/ck_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_top_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_top_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_bottom_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_bottom_2024_metat.csv\n",
      "samtools view -F 4  ~/data/metat/mp_5/n_middle_2024_metat.bam | awk 'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}' > ~/data/metat/mp_5/n_middle_2024_metat.csv\n"
     ]
    }
   ],
   "source": [
    "target_names = [f'mp_{i + 1}' for i in range(5)]\n",
    "metat_sample_ids = sample_metadata_df[sample_metadata_df.sample_id.str.contains('metat')].sample_id.values\n",
    "\n",
    "for target_name, sample_id in itertools.product(target_names, metat_sample_ids):\n",
    "    awk_pattern = '\\'BEGIN{print \"read_id,contig_id,start\";OFS=\",\"} {print $1, $4, $5}\\''\n",
    "    input_path = f'~/data/metat/{target_name}/{sample_id}.bam'\n",
    "    output_path = f'~/data/metat/{target_name}/{sample_id}.csv'\n",
    "    cmd = f'samtools view -F 4  {input_path} | awk ' + awk_pattern + f' > {output_path}'\n",
    "    print(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "57f73f40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ck_bottom_2025_metat', 'ck_middle_2025_metat',\n",
       "       'ck_top_2025_metat', 'n_top_2025_metat', 'n_bottom_2025_metat',\n",
       "       'n_middle_2025_metat', 'ck_bottom_2024_metat', 'n_top_2024_metat',\n",
       "       'n_bottom_2024_metat', 'n_middle_2024_metat'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metat_sample_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13d796dd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
