{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b01ae56f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob \n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "import os \n",
    "from utils import * \n",
    "import subprocess\n",
    "import itertools\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "798d7707",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_metadata_df = pd.read_csv('sample_metadata.csv', index_col=0)\n",
    "sample_metadata_df['metat'] = sample_metadata_df.sample_id.str.contains('metat')\n",
    "\n",
    "ggkbase_name_to_year_map = sample_metadata_df.set_index('ggkbase_name').year.to_dict()\n",
    "sample_id_to_year_map = sample_metadata_df.set_index('sample_id').year.to_dict()\n",
    "ggkbase_name_to_sample_id_map = sample_metadata_df.set_index('ggkbase_name').sample_id.to_dict()\n",
    "sample_id_to_ggkbase_name_map = sample_metadata_df.set_index('sample_id').ggkbase_name.to_dict()\n",
    "\n",
    "sample_path_template = '/groups/banfield/sequences/{year}/{ggkbase_name}/raw.d/{ggkbase_name}_trim_clean.{paired_end}.fastq.gz'\n",
    "ref_genome_dir = '../data/'\n",
    "ref_genome_paths = [f'../data/data/{file_name}.fn' for file_name in id_to_ggkbase_name_map.keys()]\n",
    "\n",
    "sample_ids = sample_metadata_df.sample_id.unique()\n",
    "genome_ids = list(id_to_ggkbase_name_map.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4748555",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_fasta_file(path:str):\n",
    "    '''Remove extra information from the FASTA file headers.'''\n",
    "    subprocess.run(f\"sed -i 's/ .*//' {path}\", shell=True, check=True)\n",
    "\n",
    "for path in ref_genome_paths:\n",
    "    clean_fasta_file(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d91d6fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "coverm_sample_paths = '' \n",
    "for row in sample_metadata_df[~sample_metadata_df.metat].itertuples():\n",
    "    coverm_sample_paths += sample_path_template.format(ggkbase_name=row.ggkbase_name, paired_end='PE.1', year=row.year) + ' '\n",
    "    coverm_sample_paths += sample_path_template.format(ggkbase_name=row.ggkbase_name, paired_end='PE.2', year=row.year) + ' '\n",
    "coverm_sample_paths = coverm_sample_paths.strip()\n",
    "\n",
    "coverm_fields = 'mean trimmed_mean covered_bases variance count rpkm tpm'\n",
    "\n",
    "with open('../scripts/coverm_mapping.sh', 'w') as f:\n",
    "    for genome_id in id_to_ggkbase_name_map.keys():\n",
    "        ref_genome_path = os.path.join(ref_genome_dir, f'{genome_id}.fn')\n",
    "\n",
    "        if 'metat' in genome_id: # Don't do this for the transcripts. \n",
    "            continue \n",
    "\n",
    "        output_file_name = f'{genome_id}.tsv'.lower()\n",
    "        output_path = f'/home/philippar/data/coverm/{output_file_name}'\n",
    "        cmd = f'coverm contig -c {coverm_sample_paths} -r {ref_genome_path} --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m {coverm_fields} -t 20 -o {output_path}'\n",
    "        f.write(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b4da042",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('../scripts/bbduk_library_sizes.sh', 'w') as f:\n",
    "#     for name, sample_id in sample_id_map.items():\n",
    "#             paired_ends_1_path = sample_path_template.format(paired_end='PE.1', sample_id=sample_id, year=sample_year_map.get(sample_id, 2025))\n",
    "#             paired_ends_2_path = sample_path_template.format(paired_end='PE.2', sample_id=sample_id, year=sample_year_map.get(sample_id, 2025))\n",
    "#             output_path = f'/home/philippar/data/bbduk/{name}.txt'\n",
    "#             cmd = f'bbduk.sh in={paired_ends_1_path} in2={paired_ends_2_path} out=/dev/null stats={output_path} threads=64'\n",
    "#             f.write(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3c84ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bbmap_get_command(sample_id:str, genome_id:str, output_dir:str='../data/metat/', input_dir='../data/'):\n",
    "    ref_genome_path = os.path.join(input_dir, f'{genome_id}.fn')\n",
    "\n",
    "    input_path_1 = sample_path_template.format(ggkbase_name=sample_id_to_ggkbase_name_map[sample_id], paired_end='PE.1', year=sample_id_to_year_map[sample_id])\n",
    "    input_path_2 = sample_path_template.format(ggkbase_name=sample_id_to_ggkbase_name_map[sample_id], paired_end='PE.2', year=sample_id_to_year_map[sample_id])\n",
    "\n",
    "    output_dir = os.path.join(output_dir, genome_id)\n",
    "    output_path = os.path.join(output_dir, f'{sample_id}.bam')\n",
    "\n",
    "    params = 'pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam'\n",
    "    cmd = f'bbmap.sh {params} in1={input_path_1} in2={input_path_2} ref={ref_genome_path} nodisk | shrinksam | sambam > {output_path}'\n",
    "    return cmd, output_path\n",
    "\n",
    "\n",
    "output_dir = '../data/metat'\n",
    "\n",
    "with open('../scripts/bbmap_mapping.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        f.write(f'mkdir -p {os.path.join(output_dir, genome_id)}\\n') # Make sure the output directory exists.\n",
    "        for sample_id in sample_ids:\n",
    "            if 'metat' in sample_id:\n",
    "                cmd, output_path = bbmap_get_command(sample_id, genome_id, output_dir=output_dir)\n",
    "                if not os.path.exists(output_path):\n",
    "                    f.write(f'sbatch --wrap \"{cmd}\"\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "185f49b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def featurecounts_get_command(bam_path:str, ref_path:str=None, output_dir:str=None):\n",
    "    output_file_name = os.path.basename(bam_path).replace('.bam', '')\n",
    "    output_file_name += '_read_counts'\n",
    "    output_path = os.path.join(output_dir, output_file_name)\n",
    "    return f'featureCounts -p -T 64 -g ID -t CDS -a {ref_path} -s 2 -o {output_path} {bam_path}' \n",
    "\n",
    "data_dir = '~/data/'\n",
    "output_dir = '~/data/metat'\n",
    "\n",
    "with open('../scripts/featurecounts_counting.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        dir_ = os.path.join(output_dir, genome_id)\n",
    "        for sample_id in sample_ids:\n",
    "            if 'metat' not in sample_id:\n",
    "                continue\n",
    "            bam_path = os.path.join(dir_, f'{sample_id}.bam')\n",
    "            ref_path = os.path.join(data_dir, f'{genome_id}.gff')\n",
    "            cmd = featurecounts_get_command(bam_path, ref_path, output_dir=dir_)\n",
    "            \n",
    "            f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3787da23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ck_bottom_2025', 'ck_middle_2025', 'ck_top_2025', 'n_bottom_2025',\n",
       "       'n_middle_2025', 'n_top_2025', 'n_middle_2024', 'n_top_2024',\n",
       "       'n_bottom_2024', 'ck_bottom_2024', '90cm_2020', '100cm_2020',\n",
       "       '100cm_2021_1', '100cm_2021_2', '80cm_2023', '40cm_2020',\n",
       "       '100cm_2022', '90cm_2023', '60cm_2023', 'ck_bottom_2025_metat',\n",
       "       'ck_middle_2025_metat', 'ck_top_2025_metat', 'n_top_2025_metat',\n",
       "       'n_bottom_2025_metat', 'n_middle_2025_metat',\n",
       "       'ck_bottom_2024_metat', 'n_top_2024_metat', 'n_bottom_2024_metat',\n",
       "       'n_middle_2024_metat'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8a099aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def interproscan_get_command(input_path:str, output_dir:str='~/data/interproscan/'):\n",
    "    id_ = os.path.basename(input_path).split('.')[0]\n",
    "    output_path = os.path.join(output_dir, f'{id_}.tsv')\n",
    "    cmd = f'interproscan.sh -i {input_path} -o {output_path} -f tsv'\n",
    "    return cmd\n",
    "\n",
    "with open('../scripts/interproscan_annotation.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        input_path = f'~/data/{genome_id}.fa'\n",
    "        cmd = interproscan_get_command(input_path)\n",
    "        f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3098b1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kofamscan_get_command(input_path:str, output_dir:str='~/data/kofamscan/'):\n",
    "    id_ = os.path.basename(input_path).split('.')[0]\n",
    "    output_path = os.path.join(output_dir, f'{id_}.tsv')\n",
    "    profiles_path = '/shared/db/kegg/kofam/latest/profiles'\n",
    "    ko_list_path = '/shared/db/kegg/kofam/latest/metadata/ko_list'\n",
    "    cmd = f'/shared/software/kofamscan/latest/exec_annotation {input_path} -o {output_path} -f detail-tsv -E 0.1 --cpu 16 -k {ko_list_path} -p {profiles_path}'\n",
    "    cmd = f'sbatch --wrap \"{cmd}\"'\n",
    "    return cmd\n",
    "\n",
    "with open('../scripts/kofamscan_annotation.sh', 'w') as f:\n",
    "    for genome_id in genome_ids:\n",
    "        input_path = f'~/data/{genome_id}.fa'\n",
    "        cmd = kofamscan_get_command(input_path)\n",
    "        f.write(cmd + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672fbe9b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "banfield",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
