{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b01ae56f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "798d7707",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_name_map = {\n",
    "    'ck_bottom_2025': 'SR-VP_Bioreactor_ck_bot_05_17_2025',\n",
    "    'ck_middle_2025': 'SR-VP_Bioreactor_ck_mid_05_17_2025',\n",
    "    'ck_top_2025': 'SR-VP_Bioreactor_ck_top_05_17_2025',\n",
    "    'n_bottom_2025': 'SR-VP_Bioreactor_N_bot_05_17_2025',\n",
    "    'n_middle_2025': 'SR-VP_Bioreactor_N_mid_05_17_2025',\n",
    "    'n_top_2025': 'SR-VP_Bioreactor_N_top_05_17_2025',\n",
    "\n",
    "    'n_middle_2024': 'SR-VP_05_06_2024_N_middle',\n",
    "    'n_top_2024':    'SR-VP_05_06_2024_N_top',   \n",
    "    'n_bottom_2024': 'SR-VP_05_06_2024_N_bottom',\n",
    "    'ck_bottom_2024': 'SR-VP_05_06_2024_ck_bottom',\n",
    "\n",
    "    '90cm_2020':  'SR-VP_26_10_2019_1_90cm',\n",
    "    '100cm_2020': 'SR-VP_26_10_2019_1_100cm',\n",
    "    '100cm_2021_1': 'SR-VP_26_10_2020_1_100CM',\n",
    "    '100cm_2021_2': 'SR-VP_26_10_2020_2_100CM',\n",
    "    '80cm_2023': 'SR-VP_11_27_2022_S1_80cm_MG',\n",
    "    '40cm_2020': 'SR-VP_26_10_2019_C_40cm',\n",
    "    '100cm_2022': 'SR-VP_11_27_2022_S1_100cm_MG',\n",
    "    '90cm_2023': 'SR-VP_07_25_2022_A1_90cm_MG',\n",
    "    '90cm_2020': 'SR-VP_26_10_2019_2_90cm',\n",
    "    '60cm_2023': 'SR-VP_11_27_2022_S1_60cm_MG',\n",
    "\n",
    "    'ck_bottom_2025_metat': 'SR-VP_Bioreactor_ck_bot_05_17_2025_metaT',\n",
    "    'ck_middle_2025_metat': 'SR-VP_Bioreactor_ck_mid_05_17_2025_metaT',\n",
    "    'ck_top_2025_metat': 'SR-VP_Bioreactor_ck_top_05_17_2025_metaT',\n",
    "    'n_top_2025_metat': 'SR-VP_Bioreactor_N_top_05_17_2025_metaT',\n",
    "    'n_bottom_2025_metat': 'SR-VP_Bioreactor_N_bot_05_17_2025_metaT',\n",
    "    'n_middle_2025_metat': 'SR-VP_Bioreactor_N_mid_05_17_2025_metaT',\n",
    "\n",
    "    'ck_bottom_2024_metat': 'SR-VP_Bioreactor_ck_bot_05_06_2024_metaT',\n",
    "    'n_top_2024_metat': 'SR-VP_Bioreactor_N_top_05_06_2024_metaT',\n",
    "    'n_bottom_2024_metat': 'SR-VP_Bioreactor_N_bot_05_06_2024_metaT',\n",
    "    'n_middle_2024_metat': 'SR-VP_Bioreactor_N_mid_05_06_2024_metaT'}\n",
    "\n",
    "\n",
    "sample_year_map = {\n",
    "    'SR-VP_26_10_2019_1_90cm': 2020,\n",
    "    'SR-VP_26_10_2019_1_100cm': 2020,\n",
    "    'SR-VP_26_10_2020_1_100CM': 2021,\n",
    "    'SR-VP_26_10_2020_2_100CM': 2021,\n",
    "    'SR-VP_11_27_2022_S1_80cm_MG': 2023,\n",
    "    'SR-VP_26_10_2019_C_40cm': 2020,\n",
    "    'SR-VP_11_27_2022_S1_100cm_MG': 2023,\n",
    "    'SR-VP_07_25_2022_A1_90cm_MG': 2023,\n",
    "    'SR-VP_26_10_2019_2_90cm': 2020,\n",
    "    'SR-VP_11_27_2022_S1_60cm_MG': 2023}\n",
    "\n",
    "input_path_template = '/groups/banfield/sequences/{year}/{sample_name}/raw.d/{sample_name}_trim_clean.{paired_end}.fastq.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91d6fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "coverm_ref_paths = ['../data/ece_26_1334.fn', '../data/methanoperedens_2.fn']\n",
    "coverm_fields = 'mean trimmed_mean covered_bases variance count rpkm tpm'\n",
    "\n",
    "# variance: Coverage variance, which is a statistical measure of how uneven the read depth is across a contig or genome.\n",
    "# trimmed_mean: Trimmed mean coverage per contig, which is essentially the mean depth after removing the lowest and highest 10% of per-base coverage values.\n",
    "#   Read depth is defined per nucleotide base. \n",
    "# rpkm: Reads per kb per million mapped reads.\n",
    "# tpm: TPM-normalized coverage, which is the RPKM for a specific contig divided by the sum of all RPKM. This controls for both contig length and sequencing depth. \n",
    "\n",
    "\n",
    "with open('../scripts/coverm_mapping.sh', 'w') as f:\n",
    "    for ref_path in coverm_ref_paths:\n",
    "        target_name = os.path.basename(ref_path).replace('.fn', '')\n",
    "        for name, sample_name in sample_name_map.items():\n",
    "            if 'metat' in name: # Don't do this for the transcripts. \n",
    "                continue \n",
    "            paired_ends_1_path = input_path_template.format(paired_end='PE.1', sample_name=sample_name, year=sample_year_map.get(sample_name, 2025))\n",
    "            paired_ends_2_path = input_path_template.format(paired_end='PE.2', sample_name=sample_name, year=sample_year_map.get(sample_name, 2025))\n",
    "            output_file_name = f'{name}-{target_name}.tsv'.lower()\n",
    "            output_path = f'/home/philippar/data/coverm/{output_file_name}'\n",
    "            cmd = f'coverm contig -1 {paired_ends_1_path} -2 {paired_ends_1_path} -r {ref_path} --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m {coverm_fields} -t 20 -o {output_path}'\n",
    "            f.write(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b4da042",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../scripts/bbduk_library_sizes.sh', 'w') as f:\n",
    "    for name, sample_name in sample_name_map.items():\n",
    "            paired_ends_1_path = input_path_template.format(paired_end='PE.1', sample_name=sample_name, year=sample_year_map.get(sample_name, 2025))\n",
    "            paired_ends_2_path = input_path_template.format(paired_end='PE.2', sample_name=sample_name, year=sample_year_map.get(sample_name, 2025))\n",
    "            output_path = f'/home/philippar/data/bbduk/{name}.txt'\n",
    "            cmd = f'bbduk.sh in={paired_ends_1_path} in2={paired_ends_2_path} out=/dev/null stats={output_path} threads=64'\n",
    "            f.write(f'sbatch --wrap \"{cmd}\"\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "360fe6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path_template = '/groups/banfield/sequences/2025/{sample_name}/raw.d/{sample_name}_trim_clean.{paired_end}.fastq.gz'\n",
    "\n",
    "ece_id = 'ece_26_1334'\n",
    "ref_paths = ['/home/philippar/data/methanoperedens_1.fn', '/home/philippar/data/methanoperedens_2.fn', f'/home/philippar/data/{ece_id}.fn']\n",
    "output_dir = '/home/philippar/data/metat/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3c84ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_mapping_command(sample_path:str, ref_path:str=None, output_dir:str=output_dir):\n",
    "    # TODO: Should look into what paired-end reads are and how that works experimentally. \n",
    "\n",
    "    target_name = os.path.basename(ref_path).replace('.fn', '')\n",
    "    sample_name = os.path.basename(sample_path)\n",
    "    input_path_1 = os.path.join(sample_path, 'raw.d', f'{sample_name}_trim_clean.PE.1.fastq.gz')\n",
    "    input_path_2 = os.path.join(sample_path, 'raw.d', f'{sample_name}_trim_clean.PE.2.fastq.gz')\n",
    "    output_path = os.path.join(output_dir, f'{sample_name_map[sample_name]}-{target_name}.bam')\n",
    "\n",
    "    params = 'pigz=t unpigz=t ambiguous=random minid=0.96 idfilter=0.97 threads=64 out=stdout.sam editfilter=5 out=stdout.sam'\n",
    "    cmd = f'bbmap.sh {params} in1={input_path_1} in2={input_path_2} ref={ref_path} nodisk | shrinksam | sambam > {output_path}'\n",
    "    return cmd, output_path\n",
    "\n",
    "\n",
    "def get_counting_command(bam_path:str, ref_path:str=None, output_dir:str=output_dir):\n",
    "    output_file_name = os.path.basename(bam_path).replace('.bam', '')\n",
    "    output_file_name += '_read_counts'\n",
    "    output_path = os.path.join(output_dir, output_file_name)\n",
    "    return f'featureCounts -p -T 64 -g ID -t CDS -a {ref_path} -s 2 -o {output_path} {bam_path}' \n",
    "\n",
    "\n",
    "def get_sbatch_command(cmd, job_name:str=None):\n",
    "    return f'sbatch --wrap \"{cmd}\" --output ../slurm.out/{job_name}.out'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f923d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0 \n",
    "\n",
    "mapping_path, counting_path = '../scripts/metat_mapping.sh', '../scripts/metat_counting.sh'\n",
    "mapping, counting = list(), list()\n",
    "for ref_path in ref_paths:\n",
    "    for sample_path in sample_paths:\n",
    "        mapping_command, bam_path = get_mapping_command(sample_path, ref_path=ref_path)\n",
    "        counting_command = get_counting_command(bam_path, ref_path=ref_path.replace('fn', 'gff'))\n",
    "        mapping.append(get_sbatch_command(mapping_command, job_name=i))\n",
    "        counting.append(counting_command)\n",
    "\n",
    "        i += 1\n",
    "\n",
    "with open(mapping_path, 'w') as f:\n",
    "    f.write('\\n'.join(mapping))\n",
    "\n",
    "with open(counting_path, 'w') as f:\n",
    "    f.write('\\n'.join(counting))"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
