sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_bottom_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_middle_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_top_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_bottom_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_middle_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_top_2025-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_middle_2024-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_top_2024-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_bottom_2024-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_bottom_2024-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/90cm_2020-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2020-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2021_1-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2021_2-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/80cm_2023-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/40cm_2020-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2022-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/90cm_2023-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.1.fastq.gz -r ../data/ece_26_1334.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/60cm_2023-ece_26_1334.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_bot_05_17_2025/raw.d/SR-VP_Bioreactor_ck_bot_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_bottom_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_mid_05_17_2025/raw.d/SR-VP_Bioreactor_ck_mid_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_middle_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_ck_top_05_17_2025/raw.d/SR-VP_Bioreactor_ck_top_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_top_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_bot_05_17_2025/raw.d/SR-VP_Bioreactor_N_bot_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_bottom_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_mid_05_17_2025/raw.d/SR-VP_Bioreactor_N_mid_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_middle_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_Bioreactor_N_top_05_17_2025/raw.d/SR-VP_Bioreactor_N_top_05_17_2025_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_top_2025-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_middle_2024-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_middle/raw.d/SR-VP_05_06_2024_N_middle_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_top_2024-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_bottom/raw.d/SR-VP_05_06_2024_N_bottom_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/n_bottom_2024-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2025/SR-VP_05_06_2024_N_top/raw.d/SR-VP_05_06_2024_N_top_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/ck_bottom_2024-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_2_90cm/raw.d/SR-VP_26_10_2019_2_90cm_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/90cm_2020-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_1_100cm/raw.d/SR-VP_26_10_2019_1_100cm_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2020-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2021/SR-VP_26_10_2020_1_100CM/raw.d/SR-VP_26_10_2020_1_100CM_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2021_1-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2021/SR-VP_26_10_2020_2_100CM/raw.d/SR-VP_26_10_2020_2_100CM_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2021_2-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_80cm_MG/raw.d/SR-VP_11_27_2022_S1_80cm_MG_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/80cm_2023-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2020/SR-VP_26_10_2019_C_40cm/raw.d/SR-VP_26_10_2019_C_40cm_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/40cm_2020-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_100cm_MG/raw.d/SR-VP_11_27_2022_S1_100cm_MG_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/100cm_2022-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_07_25_2022_A1_90cm_MG/raw.d/SR-VP_07_25_2022_A1_90cm_MG_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/90cm_2023-methanoperedens_2.tsv"
sbatch --wrap "coverm contig -1 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.1.fastq.gz -2 /groups/banfield/sequences/2023/SR-VP_11_27_2022_S1_60cm_MG/raw.d/SR-VP_11_27_2022_S1_60cm_MG_trim_clean.PE.1.fastq.gz -r ../data/methanoperedens_2.fn --min-read-percent-identity 97 --min-read-aligned-percent 80 --trim-min 5 --trim-max 95 -m mean trimmed_mean covered_bases variance count rpkm tpm -t 20 -o ../data/coverm/60cm_2023-methanoperedens_2.tsv"
